File: backend/validator/ConceptoValidator.php
<?php
// backend/validators/ConceptoValidator.php

require_once ROOT . '/backend/commons/ValidatorBase.php';

class ConceptoValidator extends ValidatorBase
{

  // Añadir propiedades específicas de concepto
  protected $tipo;
  protected $id_icono;
  protected $desembolso_planejado;
  protected $periodo_tipo;
  protected $limite_monto;
  protected $limite_tipo;

  public function __construct($data = [])
  {
    parent::__construct($data);
    $this->tipo = $data['tipo'] ?? '';
    $this->id_icono = $data['id_icono'] ?? 0;
    $this->desembolso_planejado = $data['desembolso_planejado'] ?? 0;
    $this->periodo_tipo = $data['periodo_tipo'] ?? '';
    $this->limite_monto = $data['limite_monto'] ?? 0;
    $this->limite_tipo = $data['limite_tipo'] ?? '';
  }

  public static function validate($postData): array
  {
    $validator = new self($postData);
    $errors = [];
    $cleanData = [];

    $conceptoFields = [
      'nombre' => trim($postData['nombre'] ?? ''),
      'tipo' => trim($postData['tipo'] ?? ''),
      'color' => trim($postData['color'] ?? ''),
      'id_icono' => $postData['id_icono'] ?? 0
    ];

    // Validar nombre
    if (!$validator->isValidNombreConcepto($conceptoFields['nombre'])) {
      $errors['nombre'] = $validator->getErrorMessage('nombre_concepto');
    } else {
      $cleanData['nombre'] = $validator->sanitize($conceptoFields['nombre']);
    }

    // Validar tipo
    if (!$validator->isValidTipoConcepto($conceptoFields['tipo'])) {
      $errors['tipo'] = $validator->getErrorMessage('tipo');
    } else {
      $cleanData['tipo'] = strtolower($conceptoFields['tipo']);
    }

    // Validar color (opcional)
    if (!empty($conceptoFields['color']) && !$validator->isValidColor($conceptoFields['color'])) {
      $errors['color'] = $validator->getErrorMessage('color');
    } else {
      $cleanData['color'] = $conceptoFields['color'] ?: null;
    }

    // Validar icono
    if (!$validator->isValidIcono($conceptoFields['id_icono'])) {
      $errors['id_icono'] = $validator->getErrorMessage('icono');
    } else {
      $cleanData['id_icono'] = (int) $conceptoFields['id_icono'];
    }

    if (isset($postData['configuracion_usuario']) && $postData['configuracion_usuario'] === 'true') {
      $configFields = [
        'desembolso_planejado' => $postData['desembolso_planejado'] ?? 0,
        'periodo_tipo' => $postData['periodo_tipo'] ?? '',
        'limite_monto' => $postData['limite_monto'] ?? 0,
        'limite_tipo' => $postData['limite_tipo'] ?? ''
      ];

      $configResult = $validator->validateConfiguracionUsuario($configFields);
      $errors = array_merge($errors, $configResult['errors']);
      $cleanData = array_merge($cleanData, $configResult['cleanData']);
    }

    if (!empty($errors)) {
      return [
        'success' => false,
        'errors' => $errors
      ];
    }

    return [
      'success' => true,
      'cleanData' => $cleanData
    ];
  }

  // ✅ Métodos de validación específicos para conceptos
  public function isValidNombreConcepto($nombre = ''): bool
  {
    $nombreToCheck = $nombre ?: $this->nombre;
    return strlen(trim($nombreToCheck)) >= 2
      && strlen(trim($nombreToCheck)) <= 100
      && preg_match('/^[a-zA-ZáéíóúÁÉÍÓÚñÑ0-9\s\-]+$/', trim($nombreToCheck));
  }

  public function isValidTipoConcepto($tipo = ''): bool
  {
    $tipoToCheck = $tipo ?: $this->tipo;
    $tiposValidos = ['ingreso', 'gasto'];
    return in_array(strtolower(trim($tipoToCheck)), $tiposValidos);
  }

  public function isValidIcono($id_icono = 0): bool
  {
    $idToCheck = $id_icono ?: $this->id_icono;
    return is_numeric($idToCheck) && $idToCheck > 0;
  }

  // ✅ Validar configuración de usuario para el concepto
  public function validateConfiguracionUsuario($configFields): array
  {
    $errors = [];
    $cleanData = [];

    // Validar desembolso planeado
    if (!$this->isValidMonto($configFields['desembolso_planejado'])) {
      $errors['desembolso_planejado'] = $this->getErrorMessage('monto');
    } else {
      $cleanData['desembolso_planejado'] = (float) $configFields['desembolso_planejado'];
    }

    // Validar período tipo (opcional)
    if (!empty($configFields['periodo_tipo']) && !$this->isValidPeriodoTipo($configFields['periodo_tipo'])) {
      $errors['periodo_tipo'] = $this->getErrorMessage('periodo_tipo');
    } else {
      $cleanData['periodo_tipo'] = !empty($configFields['periodo_tipo']) ? $configFields['periodo_tipo'] : null;
    }

    // Validar límite monto
    if (!$this->isValidMonto($configFields['limite_monto'])) {
      $errors['limite_monto'] = $this->getErrorMessage('monto');
    } else {
      $cleanData['limite_monto'] = (float) $configFields['limite_monto'];
    }

    // Validar límite tipo (opcional)
    if (!empty($configFields['limite_tipo']) && !$this->isValidPeriodoTipo($configFields['limite_tipo'])) {
      $errors['limite_tipo'] = $this->getErrorMessage('periodo_tipo');
    } else {
      $cleanData['limite_tipo'] = !empty($configFields['limite_tipo']) ? $configFields['limite_tipo'] : null;
    }

    return [
      'errors' => $errors,
      'cleanData' => $cleanData
    ];
  }

  public function isValidMonto($monto): bool
  {
    return is_numeric($monto) && $monto >= 0;
  }

  public function isValidPeriodoTipo($periodo_tipo): bool
  {
    $tiposValidos = ['mensual', 'quincenal', 'semanal', 'diario'];
    return in_array(strtolower(trim($periodo_tipo)), $tiposValidos);
  }

  protected function getErrorMessage($field): string
  {
    $messages = [
      'nombre_concepto' => 'El nombre del concepto debe tener entre 2 y 100 caracteres y solo puede contener letras, números y espacios',
      'tipo' => 'El tipo debe ser "ingreso" o "gasto"',
      'color' => 'El color debe estar en formato hexadecimal (#RRGGBB)',
      'icono' => 'El icono seleccionado no es válido',
      'monto' => 'El monto debe ser un número mayor o igual a 0',
      'periodo_tipo' => 'El tipo de período debe ser: mensual, quincenal, semanal o diario'
    ];

    return $messages[$field] ?? parent::getErrorMessage($field);
  }

  protected function sanitizeField($field, $value)
  {
    switch ($field) {
      case 'desembolso_planejado':
      case 'limite_monto':
        return (float) $value;
      case 'id_icono':
        return (int) $value;
      case 'periodo_tipo':
      case 'limite_tipo':
        return !empty($value) ? strtolower(trim($value)) : null;
      default:
        return parent::sanitizeField($field, $value);
    }
  }
}

File: validators/concepto_validator.go
package validators

import (
	"net/http"
	"regexp"
	"strconv"
	"strings"
)

type ConceptoValidator struct {
	*ValidatorBase
}

var ConceptoValidatorInstance = &ConceptoValidator{}

func (v *ConceptoValidator) Validate(r *http.Request) ValidationResult {
	validator := &ValidatorBase{
		Nombre: strings.TrimSpace(r.FormValue("nombre")),
		Color:  strings.TrimSpace(r.FormValue("color")),
	}

	fieldsToValidate := map[string]string{
		"nombre": validator.Nombre,
		"color":  validator.Color,
	}

	result := validator.ValidateFields(fieldsToValidate)

	// ✅ Validaciones específicas para concepto
	v.validateConceptoFields(r, &result)

	return result
}

// validateConceptoFields valida campos específicos de concepto
func (v *ConceptoValidator) validateConceptoFields(r *http.Request, result *ValidationResult) {
	// Validar tipo (gasto/ingreso)

	tipo := strings.TrimSpace(r.FormValue("tipo"))
	if tipo == "" {
		result.Errors["tipo"] = "El tipo de concepto es requerido"
		result.Success = false
	} else if tipo != "gasto" && tipo != "ingreso" {
		result.Errors["tipo"] = "El tipo debe ser 'gasto' o 'ingreso'"
		result.Success = false
	} else {
		result.CleanData["tipo"] = tipo
	}

	// Validar id_icono
	idIcono := strings.TrimSpace(r.FormValue("id_icono"))
	if idIcono == "" {
		result.Errors["id_icono"] = "Debe seleccionar un ícono"
		result.Success = false
	} else {
		iconoID, err := strconv.Atoi(idIcono)
		if err != nil || iconoID <= 0 {
			result.Errors["id_icono"] = "El ícono seleccionado no es válido"
			result.Success = false
		} else {
			result.CleanData["id_icono"] = iconoID
		}
	}

	// Validar nombre (longitud específica para BD)
	nombre := strings.TrimSpace(r.FormValue("nombre"))
	if len(nombre) > 40 {
		result.Errors["nombre"] = "El nombre no puede tener más de 40 caracteres"
		result.Success = false
	}

	// Validar configuracion_usuario
	configUsuario := strings.TrimSpace(r.FormValue("configuracion_usuario"))
	if configUsuario != "true" && configUsuario != "false" {
		result.Errors["configuracion_usuario"] = "Configuración de usuario inválida"
		result.Success = false
	} else {
		result.CleanData["configuracion_usuario"] = configUsuario == "true"
	}

	// ✅ Validar campos de personalización (opcionales)
	v.validatePersonalizacion(r, result)

	// Si hay errores, actualizar el estado
	if len(result.Errors) > 0 {
		result.Success = false
	}
}

// validatePersonalizacion valida los campos opcionales de personalización
func (v *ConceptoValidator) validatePersonalizacion(r *http.Request, result *ValidationResult) {
	// Desembolso planificado
	desembolsoStr := strings.TrimSpace(r.FormValue("desembolso_planejado"))
	if desembolsoStr != "" {
		desembolso, err := strconv.ParseFloat(desembolsoStr, 64)
		if err != nil || desembolso < 0 {
			result.Errors["desembolso_planejado"] = "El desembolso planificado debe ser un número positivo"
			result.Success = false
		} else {
			result.CleanData["desembolso_planejado"] = desembolso
		}
	} else {
		result.CleanData["desembolso_planejado"] = 0.0
	}

	// Período tipo para desembolso
	periodoTipo := strings.TrimSpace(r.FormValue("periodo_tipo"))
	if periodoTipo != "" && periodoTipo != "diario" && periodoTipo != "quincenal" && periodoTipo != "mensual" {
		result.Errors["periodo_tipo"] = "El tipo de período debe ser: diario, quincenal o mensual"
		result.Success = false
	} else {
		result.CleanData["periodo_tipo"] = periodoTipo
	}
	// Dia Periodo tipo para desembolso
	diaDesembolsoStr := strings.TrimSpace(r.FormValue("dia_desembolso_planejado"))
	if diaDesembolsoStr != "" {
		diaDesembolso, err := strconv.ParseInt(diaDesembolsoStr, 10, 8)
		if err != nil {
			result.Errors["dia_desembolso_planejado"] = "El día debe ser un número válido (Desembolso planificado)."
			result.Success = false
		} else if diaDesembolso < 1 || diaDesembolso > 31 {
			result.Errors["dia_desembolso_planejado"] = "El día debe estar en un rango de 1 - 31"
			result.Success = false
		} else {
			// GUARDAR EL VALOR CONVERTIDO como int8
			result.CleanData["dia_desembolso_planejado"] = int8(diaDesembolso)
		}
	} else {
		// Si no hay valor, establecer nil para que no se guarde en BD
		result.CleanData["dia_desembolso_planejado"] = nil
	}

	// Validación adicional: si período es mensual, requiere día
	if periodoTipo == "mensual" && diaDesembolsoStr == "" {
		result.Errors["dia_desembolso_planejado"] = "El día es requerido para período mensual"
		result.Success = false
	}

	// Si período es quincenal, forzar día 15
	if periodoTipo == "quincenal" {
		result.CleanData["dia_desembolso_planejado"] = int8(15)
	}

	// Límite de monto
	limiteStr := strings.TrimSpace(r.FormValue("limite_monto"))
	if limiteStr != "" {
		limite, err := strconv.ParseFloat(limiteStr, 64)
		if err != nil || limite < 0 {
			result.Errors["limite_monto"] = "El límite debe ser un número positivo"
			result.Success = false
		} else {
			result.CleanData["limite_monto"] = limite
		}
	} else {
		result.CleanData["limite_monto"] = 0.0
	}

	// Límite tipo
	limiteTipo := strings.TrimSpace(r.FormValue("limite_tipo"))
	if limiteTipo != "" && limiteTipo != "diario" && limiteTipo != "quincenal" && limiteTipo != "mensual" {
		result.Errors["limite_tipo"] = "El tipo de límite debe ser: diario, quincenal o mensual"
		result.Success = false
	} else {
		result.CleanData["limite_tipo"] = limiteTipo
	}

	// Dia Periodo tipo para desembolso
	diaLimiteStr := strings.TrimSpace(r.FormValue("dia_limite_tipo"))
	if diaLimiteStr != "" {
		diaLimite, err := strconv.ParseInt(diaLimiteStr, 10, 8)
		if err != nil {
			result.Errors["dia_limite_tipo"] = "El día debe ser un número válido (Establecer límite)."
			result.Success = false
		} else if diaLimite < 1 || diaLimite > 31 {
			result.Errors["dia_limite_tipo"] = "El día debe estar en un rango de 1 - 31"
			result.Success = false
		} else {
			// ✅ GUARDAR EL VALOR CONVERTIDO como int8
			result.CleanData["dia_limite_tipo"] = int8(diaLimite)
		}
	} else {
		// Si no hay valor, establecer nil para que no se guarde en BD
		result.CleanData["dia_limite_tipo"] = nil
	}

	// Validación adicional: si límite tipo es mensual, requiere día
	if limiteTipo == "mensual" && diaLimiteStr == "" {
		result.Errors["dia_limite_tipo"] = "El día es requerido para límite mensual"
		result.Success = false
	}

	// Si límite tipo es quincenal, forzar día 15
	if limiteTipo == "quincenal" {
		result.CleanData["dia_limite_tipo"] = int8(15)
	}

	// Notificación (booleano)
	notificacionStr := strings.TrimSpace(r.FormValue("notificacion"))
	if notificacionStr != "" {
		notificacion, err := strconv.ParseBool(notificacionStr)
		if err != nil {
			result.Errors["notificacion"] = "La notificación debe ser verdadero o falso"
			result.Success = false
		} else {
			result.CleanData["notificacion"] = notificacion
		}
	} else {
		result.CleanData["notificacion"] = false
	}
}

// ValidateUpdate valida la actualización de un concepto
func (v *ConceptoValidator) ValidateUpdate(r *http.Request) ValidationResult {
	result := v.Validate(r)

	// Validaciones adicionales para actualización
	nombreConcepto := strings.TrimSpace(r.FormValue("nombre_original"))
	if nombreConcepto == "" {
		result.Errors["nombre_original"] = "El nombre original del concepto es requerido para actualizar"
		result.Success = false
	} else {
		result.CleanData["nombre_original"] = nombreConcepto
	}

	return result
}

// ValidatePersonalizacion valida solo los campos de personalización
func (v *ConceptoValidator) ValidatePersonalizacion(r *http.Request) ValidationResult {
	result := NewValidationResult()

	// Validar que el concepto existe
	nombreConcepto := strings.TrimSpace(r.FormValue("nombre_concepto"))
	correoFamilia := strings.TrimSpace(r.FormValue("correo_familia"))

	if nombreConcepto == "" {
		result.Errors["nombre_concepto"] = "El nombre del concepto es requerido"
	}
	if correoFamilia == "" {
		result.Errors["correo_familia"] = "El correo de la familia es requerido"
	}

	if len(result.Errors) > 0 {
		result.Success = false
		return result
	}

	result.CleanData["nombre_concepto"] = nombreConcepto
	result.CleanData["correo_familia"] = correoFamilia

	// Validar campos de personalización
	v.validatePersonalizacion(r, &result)

	return result
}

// ValidateNombreUnico verifica si el nombre del concepto es único para la familia
// Esta función se usaría en el controlador después de validar los datos básicos
func (v *ConceptoValidator) ValidateNombreUnico(nombreConcepto, correoFamilia string, esActualizacion bool, nombreOriginal string) (bool, string) {
	// Esta validación requiere acceso a la base de datos
	// Se implementaría en el controlador llamando al modelo

	if esActualizacion && nombreConcepto == nombreOriginal {
		return true, "" // Mismo nombre en actualización, no hay problema
	}

	// Para implementación completa, necesitarías inyectar el modelo o hacer esta validación en el controlador
	return true, "" // Placeholder
}

// Helper para validar formato de ícono (FontAwesome)
func (v *ConceptoValidator) IsValidIcono(icono string) bool {
	if icono == "" {
		return false
	}

	// Validar formato básico de clase FontAwesome
	// Ej: "fa-solid fa-house", "fa-regular fa-user", etc.
	iconoRegex := `^fa-[a-z]+(-[a-z]+)* fa-[a-z]+(-[a-z]+)*$`
	matched, _ := regexp.MatchString(iconoRegex, strings.TrimSpace(icono))
	return matched
}
